### ISSUE TITLE
Periodic test failures in GitHub Actions CI

### ISSUE THREAD

**Issue by jywarren** [publiclab/plots2#9873]
I'm seeing occasional errors that seem to self-resolve when we re-run the tests. I'm not sure the cause. Here's one log I found:

```
2021-06-29 20:46:08 +0000 Rack app ("GET /barnstar/givestar=basic&nid=1" - (127.0.0.1)): Mysql2::Error
=[Screenshot]: tmp/screenshots/failures_test_awarding_barnstar_functions_correctly.png
ERROR PostTest#test_awarding_barnstar_functions_correctly (873.23s)
Minitest::UnexpectedError:         Mysql2::Error: 
            app/models/tag.rb:430:in `update_tags_activity'
            app/models/node.rb:678:in `tag_activity'
            app/models/node.rb:671:in `add_comment'
            app/controllers/tag_controller.rb:284:in `barnstar'
```

These can sometimes happen due to timing issues -- because tests run all at once with no time passing in between. We can sometimes resolve them by adding an order clause to the SQL, or by using TimeCop to manipulate the time during tests. Let's collect more data to see what tests are inconsistent (paste them here before restarting them, please) and go from there. cc @publiclab/soc

**Comment by jywarren** [comment]
Confirmed the exact same bug blocked https://github.com/publiclab/plots2/pull/9848 - and is blocking lots of dependabot PRs too, so that's a good place to look for this happening!

**Comment by jywarren** [comment]
https://github.com/publiclab/plots2/pull/9864/ as well! So, what is the actual error? 

It seems related to this line from https://github.com/publiclab/plots2/pull/9165 by @RuthNjeri --

https://github.com/publiclab/plots2/blob/78bf29f75865cb6d0dd5e5f55e3ff24ea707a65a/app/models/tag.rb#L430

No worries, Ruth, it seems super obscure. But we're doing an `update_all` -- could it be blocked by some other database lock? We are running tests in parallel, which kind of simulates how this could run in production with many users at once, you know? 

Could we do this differently?

**Comment by daemon1024** [comment]
I wonder if introducing pessimistic locking will help us prevent this race condition. 

https://api.rubyonrails.org/classes/ActiveRecord/Locking/Pessimistic.html

**Comment by jywarren** [comment]
Oh, interesting! So how would that look? Like this?

```ruby
Tag.transaction do
 Tag.lock
  .where(tid: tids)
  .update_all(activity_timestamp: DateTime.now, latest_activity_nid: activity_id) 
end
```

@RuthNjeri @cesswairimu , what do you think? I've never done this before! 😅

**Comment by jywarren** [comment]
@icarito, would it be OK to lock some tag records in this way while the rows are being updated together? Or would it cause some adverse performance? 

I'm wondering if:

1. we should only lock in test mode, to avoid a performance issue in production, or
2. we should just lock in general because we may have this kind of MySQL error happening in production?

We should be able to check for such errors in Sentry. I don't think i've been seeing any MySQL related errors though, so i think this should be OK to leave as-is in production mode.

**Comment by jywarren** [comment]
For reference, I think we're talking about locking about 6 rows in the `term_data` table (the tags table) while we update those 6 rows. So not a huge issue, and only when tags are added to a node:

See https://github.com/publiclab/plots2/pull/9165/files#diff-2845087add773fb20ddff83e51ab1ef70788ad7e522c7363e03458c6adab93b3R11

**Comment by jywarren** [comment]
I've attempted the fix in https://github.com/publiclab/plots2/pull/9881 so let's see what happens there.

**Comment by cesswairimu** [comment]
> Oh, interesting! So how would that look? Like this?
> 
> ```ruby
> Tag.transaction do
>  Tag.lock
>   .where(tid: tids)
>   .update_all(activity_timestamp: DateTime.now, latest_activity_nid: activity_id) 
> end
> ```
> 
> @RuthNjeri @cesswairimu , what do you think? I've never done this before!

This is looking good @jywarren :tada:

**Comment by RuthNjeri** [comment]
Hi @jywarren and everyone 👋🏾 
I'm having a look at this now, it seems interesting locking the database transaction... I thought the fact that MySQL is relational, it will lock a transaction based on the ACID property...

If the tag locking does not work, we could look into mocking this for the tests...

**Comment by jywarren** [comment]
OK, we merged https://github.com/publiclab/plots2/pull/9881 to try to address this. Let's watch out and i suspect (and @Manasa2850 noted something related to this too) that there's another issue sometimes causing test failures. We can close it if things have stabilized in a week or two!

I did notice that CodeCov failed twice for @17sushmita in https://github.com/publiclab/plots2/pull/9885 and https://github.com/publiclab/plots2/pull/9879 for no apparent reason, misreading test coverage changes. So let's watch that too, and if you see that behavior please mention and link us to it here! Thanks, all!

**Comment by jywarren** [comment]
Lots of codecov failures w dependabot: https://github.com/publiclab/plots2/pull/9887

**Comment by jywarren** [comment]
See here for example I asked dependabot to rebase https://github.com/publiclab/plots2/pull/9886#issuecomment-874962061

**Comment by Tlazypanda** [comment]
So I might be shooting in the dark but I found this issue https://github.com/codecov/codecov-action/issues/330 and this link https://stackoverflow.com/questions/67861379/codecov-fails-in-github-actions which suggests moving to 1.5.2 should solve our issue and then in our `dependabot.yml` file `codecov` dependency is under `ignore` category so could it be that we aren't updating our codecov and we are facing this? not sure if it gets updated in another way as I said throwing darts :p @jywarren @cesswairimu

**Comment by cesswairimu** [comment]
oh nice @Tlazypanda checking this out. Thanks

**Comment by cesswairimu** [comment]
@jywarren I checked and couldn't find anywhere we could specify the codecov version. Any ideas? thanks

**Comment by jywarren** [comment]
Great digging here, @Tlazypanda @cesswairimu !!!

Digging into when that ignore got added...

#6290 relevant... also #9756 cited the same CodeCov issue!

https://github.com/publiclab/plots2/pull/9552 is maybe where CodeCov was last updated, on April 23 2021

https://github.com/publiclab/plots2/blob/6832a437ed2a6451d6bc30bb9876906a9ec92eb8/Gemfile.lock#L89-L90 shows v1.5.2 so we are already on that version...

https://github.com/publiclab/plots2/pull/9583/files is where the newer Dependabot added the ignore? 

So wait, the ignore is only to ignore that specific version, not later ones. It's generated when we close a Dependabot PR on the specific version -- it "remembers" that we don't want to update to that particular version. But we got past it so we're now on v1.5.2:

https://github.com/publiclab/plots2/blob/6832a437ed2a6451d6bc30bb9876906a9ec92eb8/.github/dependabot.yml#L43-L45

**Comment by jywarren** [comment]
Looking at a recent fail here for exact reference: https://github.com/publiclab/plots2/pull/9894

**Comment by jywarren** [comment]
So there seems to be some error for 9894 in CodeCov:

![image](https://user-images.githubusercontent.com/24359/125514854-8c9b6b49-ba87-4019-9f09-6d6fa0cdb64c.png)

**Comment by jywarren** [comment]
I'm going to do 2 things to see if we get past it:

1. read and implement https://docs.codecov.com/docs/error-reference#section-missing-base-commit to get past that issue
2. get a new token, as https://github.com/codecov/codecov-action/issues/330 recommends

**Comment by jywarren** [comment]
Also noting some people are using a specific GitHub Actions step like this:

```
    - name: Upload to codecov
      uses: codecov/codecov-action@v1
```

**Comment by jywarren** [comment]
Could be this? https://community.codecov.com/t/unable-to-determine-a-parent-commit-to-compare-against-in-base-branch-after-squash-and-merge/2480/20

**Comment by jywarren** [comment]
Oh hmm, i see this message:

>  New! Install Codecov's GitHub App to improve application link. Comment posting will be on behalf of Codecov not a bot user. Install now

Sure! I'll try that!

**Comment by jywarren** [comment]
OK i tried reinstalling the app. I thought it'd be a different "Github app" but it looked the same... in any case hopefully it'll "reset" something.

I also tried this solution for the "missing base commit" error: #9905

Also i noticed this amazing graph of the issues happening over time:

![image](https://user-images.githubusercontent.com/24359/125518979-e6c1ea07-561a-48f2-9741-9902193f37f5.png)

**Comment by jywarren** [comment]
I don't know if this is a new feature, but i saw this `rerun` link and clicked it on #9894: 

![image](https://user-images.githubusercontent.com/24359/125519414-f67e8825-cd5d-4505-b9dc-f3370f2c7684.png)

Could be a new feature of the new app, or just something i didn't notice before?

**Comment by jywarren** [comment]
OK so for what it's worth, Re-run failed in 1 second: `codecov/project Failing after 1s — 49.24% (-29.00%) compared to d12a468`

**Comment by jywarren** [comment]
https://docs.codecov.com/docs/comparing-commits

In general though, lots of people are using this: https://github.com/codecov/codecov-action

I'm not actually sure how ours is run! Is it some hook? Originally installed in https://github.com/publiclab/plots2/pull/5954

But, this looks like it may have worked??

https://github.com/publiclab/plots2/pull/9905

**Comment by jywarren** [comment]
Aha - here's where it's run in each test job: 

https://github.com/publiclab/plots2/blob/d9ad9e4d3618d5a1a440ba540f622f6dee911c96/test/test_helper.rb#L1

**Comment by jywarren** [comment]
I tweaked quite a bit in the codecov.yml file in #9905 - following also guidance here:

https://docs.codecov.com/docs/comparing-commits#pseudo-comparison

**Comment by jywarren** [comment]
OK, so this still isn't working and we are currently uploading all reports twice, even using the CodeCov GitHub Action which is the 'preferred' up-to-date way.

This suggests storing the reports as artifacts and then uploading them all together at the end: https://github.com/codecov/codecov-action/issues/300

It sounds great. Let's try that.

**Comment by jywarren** [comment]
Also noting that we aren't seeing errors like https://github.com/codecov/codecov-action/issues/330; we're just seeing inconsistent and varying coverage percentages, possibly because it's not properly integrating the reports from each job (unit tests, integration tests, etc) being uploaded at different times.

**Comment by jywarren** [comment]
Also noting that March 19 was around when we started seeing this much lower coverage reporting. And, I am seeing one instance of the barnstar test failure I thought we'd fixed, above: https://github.com/publiclab/plots2/pull/9905#issuecomment-879434128

However, it was one of like 10 test runs, so not too common.

**Comment by jywarren** [comment]
OK, attempting a "merge-pr" action which uploads new coverage after a PR is merged. If that doesn't work, we might try this template for compiling up all reports before uploading:

https://github.com/codecov/example-actions-bundled/blob/2ab79565a2e2daa4d07857899b1089437461baac/.github/workflows/ci.yml#L11-L29

**Comment by jywarren** [comment]
I think this last one where we upload a report for the base branch after each PR is promising. The low CodeCov coverage score is suspect in part because it's comparing against some random commit:

`codecov/project — 26.04% (-52.20%) compared to d12a468` where https://github.com/publiclab/plots2/commit/d12a468 is not even in our base branch tree -- it's a commit from within this squashed PR: https://github.com/publiclab/plots2/pull/9810/commits

However, the very low coverage is also mysterious. No matter what it's compared against, how could it have a project coverage percentage of 26%? So, we may need to compile the reports before uploading as well.

**Comment by jywarren** [comment]
The "wrong base for comparison" theory matches the error message I get when [viewing this PR in CodeCov](https://app.codecov.io/gh/publiclab/plots2/compare/9905):

```
Oops, something went wrong
Changes found in between d12a468...d9ad9e4 (pseudo...base) which prevent comparing this pull request. Check on Codecov's status or see our docs for common support.

Error 400
```

that pseudo base is not even in our base branch tree.

**Comment by jywarren** [comment]
Working a little on the barnstar system test timing issue which is quite rare but happened twice in #9905 over like 20 runs... 

https://github.com/publiclab/plots2/pull/9909

**Comment by jywarren** [comment]
OK, we had a slight hiccup where the merge-pr action was in the wrong folder. Moved it and adjusted in https://github.com/publiclab/plots2/pull/9910

**Comment by jywarren** [comment]
OK, hopeful this will work: https://github.com/publiclab/plots2/actions/runs/1038211403

🤞

**Comment by jywarren** [comment]
https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#permissions

**Comment by jywarren** [comment]
https://docs.github.com/en/actions/reference/authentication-in-a-workflow

**Comment by jywarren** [comment]
OMG ok this took a while. I think i got it. We now have this which triggers after a merge to main:

https://github.com/publiclab/plots2/runs/3088995766?check_suite_focus=true

That then is used as a trigger for another run of the regular tests workflow, which strangely here ran twice:

https://github.com/publiclab/plots2/actions/runs/1038480174

https://github.com/publiclab/plots2/actions/runs/1038480867

One says `jywarren requested` and the other `jywarren completed` so maybe we should trigger it on a specific event:

OK done: https://github.com/publiclab/plots2/pull/9916

**Comment by jywarren** [comment]
https://github.com/actions/runner/issues/950

**Comment by jywarren** [comment]
Hmm. it's still running 2x. But the sequence is now:

1. https://github.com/publiclab/plots2/actions/runs/1038489457 - original PR
2. https://github.com/publiclab/plots2/actions/runs/1038493942 to trigger one more run after merging PR
3. https://github.com/publiclab/plots2/actions/runs/1038493985 and https://github.com/publiclab/plots2/actions/runs/1038494629 run

This is good enough for now!

**Comment by jywarren** [comment]
Now i'll open a new PR and we'll see what it compares against!

**Comment by jywarren** [comment]
I'm also hoping we'll see "Missing base report" on this page go away:

https://app.codecov.io/gh/publiclab/plots2/pulls?page=1&state=open&order=-pullid

![image](https://user-images.githubusercontent.com/24359/125996401-960327f1-90d6-48c9-9fc2-33f256afb19c.png)

**Comment by jywarren** [comment]
Ooh! I think that's it!

![image](https://user-images.githubusercontent.com/24359/125997607-2b5fe0b8-f998-471d-af77-8c56a850bb2a.png)

**Comment by jywarren** [comment]
And https://github.com/publiclab/plots2/pull/9909 | https://app.codecov.io/gh/publiclab/plots2/compare/9909 is now not showing "Missing base report" anymore! 🎉

**Comment by jywarren** [comment]
OK - the good news is that the comparison is against the correct base commit: `compared to 3d3de3f`

However, the /check/ isn't passing, and is reporting 59.59% (-22.56%) which is not right. CodeCov itself reports 74.07% here: https://app.codecov.io/gh/publiclab/plots2/compare/9909/. I think the 59.59% is from one commit behind. I.e. before it was really ready to report back. The "re-run" link doesn't work.

**Comment by jywarren** [comment]
After looking in documentation, i finally just reported the discrepancy to CodeCov in their forums: https://community.codecov.com/t/github-status-check-reported-too-early/3043

https://docs.codecov.com/docs/github-checks

https://docs.codecov.com/docs/commit-status

**Comment by cesswairimu** [comment]
thanks Jeff :rocket: :rocket:
